### PR TITLE
lnrpc+htlcswitch+channeldb: add new forwarding log time series and query interface

### PR DESCRIPTION
**Pull request by Roasbeef**
In this commit, we add a much requested feature: the ability to track the fees successfully payment circuits have carried, and also examine the history of HTLCs successfully forwarded by the `htlcswitch`. 

## new `channeldb` storage 

In this PR, we add a new storage namespace to channeldb: the `ForwardingLog`. This log will be used by higher level sub-systems to log each successfully completed HTLC. Each payment circuit will be
summarized as a `ForwardingEvent`. A series of events can then be queried via a time slice query. In a time slice query, the caller specifies a time range, a number of events to skip, and the max number of events to return. Each query will return the index of the final item. As we have a max number of events we’ll return in a response, callers may need to use this last offset index to seek further by skipping that number of entries. Combining these fields, callers are able to query the time series, skipping an arbitrary amount of events, and capping the max number of returned events.

We store the timestamp in nanoseconds since the unix epoch. Using this level of granularity, we assume that we'll never have a repeated time stamp (due to the monotonicity of recorded time in the Go runtime). 

## new `hltcswitch` functionality

In this PR, we extend the switch as is, to record details concerning settled payment circuits. To do this, we introduce a new interface to the package: the `ForwardingLog`. This is a tiny interface that simply lets us abstract away the details of the storage backing of the forwarding log.

Each time we receive a successful HTLC settle, we’ll log the full details (chans, fees, time) as a pending forwarding log entry. Every 15 seconds, we’ll then batch flush out these entries to disk. When we’re exiting, we’ll try to flush out all entries to ensure everything gets recorded to disk.
 
## RPC Modifications 
The `FeeReport` command has been extended to include the recorded fee revenue over the past day, week, and month. We've also added a `ForwardingHistory` command that lets callers query the entire time series, skip events in the time series, and specify a max number of events to return. A new `lncli` command `fwdinghistory` has been added to expose this query information on the command line.

### CODE REVIEW

**Comment by Roasbeef** [comment]
Some sample outputs from the faucet: 
```
$ lncli feereport
<snip> 
   ],
    "day_fee_sum": "2",
    "week_fee_sum": "2",
    "month_fee_sum": "2"
}

$ lncli fwdinghistory 
{
    "forwarding_events": [
        {
            "timestamp": "1519857535",
            "chan_id_in": "1414915334303449089",
            "chan_id_out": "1414950518670229504",
            "amt_in": "1001",
            "amt_out": "1000",
            "fee": "1"
        },
        {
            "timestamp": "1519857771",
            "chan_id_in": "1414994499134226433",
            "chan_id_out": "1414305105343741953",
            "amt_in": "1211",
            "amt_out": "1210",
            "fee": "1"
        }
    ],
    "last_offset_index": 2
}
```

I have a TODO to possibly use FPE on-disk for the forwarding log...

**Comment by Roasbeef** [comment]
@aakselrod PTAL.

File: channeldb/forwarding_log.go [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+	"github.com/boltdb/bolt"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+var (
+	// forwardingLogBucket is the bucket that we'll use to store the
+	// forwarding log. The forwarding log contains a time series database
+	// of the forwarding history of a lightning daemon. Each key within the
+	// bucket is a timestamp (in nano seconds since the unix epoch), and
+	// the value a slice of a forwarding event for that timestamp.
```
Comment (aakselrod) [inline channeldb/forwarding_log.go]: If keeping a single forwarding event per timestamp, remove "slice of" here.

File: channeldb/forwarding_log.go [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+// particular time slice. The caller can control the precise time as well as
+// the number of events to be returned.
+//
+// TODO(roasbeef): rename?
+func (f *ForwardingLog) Query(q ForwardingEventQuery) (ForwardingLogTimeSlice, error) {
+	resp := ForwardingLogTimeSlice{
+		ForwardingEventQuery: q,
+	}
+
+	// If the user provided an index offset, then we'll not know how many
```
Comment (aakselrod) [inline channeldb/forwarding_log.go]: Remove "not" here

File: htlcswitch/switch_test.go [hunk 3]
```diff
@@ -542,3 +569,165 @@ func TestSwitchSendPayment(t *testing.T) {
...
+	}
+	defer cleanUp()
+
+	n := newThreeHopNetwork(t, channels.aliceToBob, channels.bobToAlice,
+		channels.bobToCarol, channels.carolToBob, testStartingHeight)
+	if err := n.start(); err != nil {
+		t.Fatalf("unable to start three hop network: %v", err)
+	}
+
+	// We'll make now 10 payments, of 100k satoshis each from Alice to
```
Comment (aakselrod) [inline htlcswitch/switch_test.go]: Maybe "now make" but this is fine.

**Pull request merged**